### ISSUE TITLE
A question on implementation

### ISSUE THREAD

**Issue by jambo6** [silvandeleemput/memcnn#71]
Hello,

Thanks a bunch for this package. I've had a go at implementing something similar following your code, I have one query about the implementation. It relates to the following lines
```
if not ctx.keep_input:
    # PyTorch 1.0+ way to clear storage
    inputs[0].storage()
    for element in inputs:
        element.storage().resize_(0)
...
# recompute input
with torch.no_grad():
    inputs_inverted = ctx.fn_inverse(*outputs)
    if not isinstance(inputs_inverted, tuple):
        inputs_inverted = (inputs_inverted,)
    for element_original, element_inverted in zip(inputs, inputs_inverted):
        element_original.storage().resize_(int(np.prod(element_original.size())))
        element_original.set_(element_inverted
```
Why do you go via this approach to free up the memory?

I thought it would have been sufficient to do something along the lines of 
```
del inputs
torch.cuda.empty_cache()
```
and then simply reconstruct with 
```
with torch.no_grad():
    inputs_inverted = ctx.fn_inverse(*outputs)
```
ignoring the `storage().resize_` lines altogether. However the second approach doesn't work, whereas your approach frees the memory.

Could you enlighten me as to what exactly is going on in the first approach, and why it is required over something more straightforward where we simply free the `inputs` memory in `forward` and reconstruct in `backwards`?

Thanks!

**Comment by silvandeleemput** [comment]
I have done a few experiments on the numerical error of the reconstructed gradients w.r.t. the normal gradients and indeed I also found that the error increased exponentially with the number of invertible layers. 


Also check figure 4 from the RevNet paper: https://arxiv.org/pdf/1707.04585.pdf. Note the section in the paper on the numerical error with a suggestion to combat it if necessary (with some additional overhead):

```
Numerical error. While Eqn. 8 reconstructs the activations exactly when done in exact arithmetic,
practical float32 implementations may accumulate numerical error during backprop. We study the
effect of numerical error in Section 5.2; while the error is noticeable in our experiments, it does not
significantly affect final performance. We note that if numerical error becomes a significant issue,
one could use fixed-point arithmetic on the x’s and y’s (but ordinary floating point to compute F and
G), analogously to [19]. In principle, this would enable exact reconstruction while introducing little
overhead, since the computation of the residual functions and their derivatives (which dominate the
computational cost) would be unchanged.
```

**Comment by jambo6** [comment]
Thanks!
